controller/array_action/array_mediator_svc.py

```diff
@@ -510,10 +511,31 @@ def copy_to_existing_volume(self, volume_id, source_id, source_capacity_in_bytes
         target_volume_name = self._get_volume_name_by_wwn(volume_id)
         self._copy_to_target_volume(target_volume_name, source_name)
 
-    def create_volume(self, name, size_in_bytes, space_efficiency, pool, io_group):
-        self._create_cli_volume(name, size_in_bytes, space_efficiency, pool, io_group)
+    def _create_cli_volume_from_snapshot(self, name, pool, io_group, source_id):
+        logger.info("creating volume from snapshot")
+        response = self._mkvolumegroup(name, pool, io_group, source_id)
+        volume_group_id = self._get_id_from_response(response)
+        volume_id = self._get_volume_id_from_volume_group(volume_group_id)
+
+        self._extract_and_remove_volume_group(volume_id, volume_group_id)
+        self._rename_volume(volume_id, name)
+
+    def _create_cli_volume_with_snapshot(self, name, pool, io_group, source_ids, source_type):
+        if source_type == controller_config.SNAPSHOT_TYPE_NAME:
+            self._create_cli_volume_from_snapshot(name, pool, io_group, source_ids.internal_id)
+
+    def _is_source_support_addsnapshot(self, object_uid):
+        return self._is_addsnapshot_supported() and not self._is_source_have_fcmaps(object_uid)
+
+    def create_volume(self, name, size_in_bytes, space_efficiency, pool, io_group, source_ids, source_type):
+        copied = False
+        if source_type and self._is_source_support_addsnapshot(source_ids.object_uid):
+            self._create_cli_volume_with_snapshot(name, pool, io_group, source_ids, source_type)
```

Comment:
```suggestion
    def _is_source_support_addsnapshot(self, object_uid):
        return self._is_addsnapshot_supported() and not self._is_source_have_fcmaps(object_uid)

    def create_volume(self, name, size_in_bytes, space_efficiency, pool, io_group, source_ids, source_type):
        copied = False
        if source_type == controller_config.SNAPSHOT_TYPE_NAME and \
           self._is_source_support_addsnapshot(source_ids.object_uid):
            self._create_cli_volume_from_snapshot(name, pool, io_group, source_ids.internal_id, source_type)
```
`with_snapshot` sounds to me like it will create a snapshot as well

Reply:
it's not from snapshot necessarily, it can be from a volume when cloning, and in this case, it will indeed create a snapshot. so in both source types, it will be using snapshot but not from snapshot.

Reply:
oh, so it's actually accurate, but the snapshot creation is not part of our code?

Reply:
it will be when I will add clone support.

Reply:
OK, I guess it will be easier to discuss this after clone support